internal/app/dogma/get_cmd.go

```diff
@@ -66,6 +78,124 @@ func (gf *getFileCommand) execute(c *cli.Context) error {
 	return nil
 }
 
+type getDirectoryCommand struct {
+	out           io.Writer
+	repo          repositoryRequestInfo
+	localFilePath string
+}
+
+func (gd *getDirectoryCommand) execute(c *cli.Context) error {
+	client, err := newDogmaClient(c, gd.repo.remoteURL)
+	if err != nil {
+		return err
+	}
+
+	return gd.executeWithDogmaClient(c, client)
+}
+
+func (gd *getDirectoryCommand) executeWithDogmaClient(_ *cli.Context, client *centraldogma.Client) error {
+	repo := gd.repo
+	entry, err := getRemoteFileEntryWithDogmaClient(client,
+		repo.projName, repo.repoName, repo.path, repo.revision, nil)
+	if err != nil {
+		return err
+	}
+
+	if entry.Type != centraldogma.Directory && repo.isRecursiveDownload {
+		return fmt.Errorf("%+q is not a directory, you might want to remove `--recursive` instead", repo.path)
+	}
+
+	basename := creatableFilePath(gd.localFilePath, 1)
+	if err := os.MkdirAll(basename, defaultPermMode); err != nil {
+		return err
+	}
+	return gd.recurseDownload(context.Background(), client, basename, entry)
+}
+
+func (gd *getDirectoryCommand) recurseDownload(ctx context.Context, client *centraldogma.Client,
+	basename string, rootEntry *centraldogma.Entry) error {
+	if rootEntry.Type != centraldogma.Directory {
+		return fmt.Errorf("%+q is not a directory, you might want to remove `--recursive` instead",
+			rootEntry.Path)
+	}
+
+	repo := gd.repo
+	path := rootEntry.Path
+	entries, httpStatusCode, err := client.ListFiles(ctx, repo.projName, repo.repoName, repo.revision, path)
+	if err != nil {
+		return err
+	}
+
+	if httpStatusCode != http.StatusOK {
+		return fmt.Errorf("failed to get the list of files in the /%s/%s%s revision: %q (status: %d)",
+			repo.projName, repo.repoName, path, repo.revision, httpStatusCode)
+	}
+
+	for _, entry := range entries {
+		switch entry.Type {
+		case centraldogma.Directory:
+			if err := gd.recurseDownload(ctx, client, basename, entry); err != nil {
+				return err
+			}
+		default:
+			if err := gd.downloadFile(client, basename, entry.Path); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (gd *getDirectoryCommand) downloadFile(client *centraldogma.Client, basename, path string) error {
+	repo := gd.repo
+	name := gd.constructFilename(basename, path)
+	if err := os.MkdirAll(filepath.Dir(name), defaultPermMode); err != nil {
+		return err
+	}
+	fd, err := os.Create(name)
+	if err != nil {
```

Comment:
I think we miss an important point here:
- On error (no matter what error is), we should remove created temporary file

In this way, how about:

```go
fd, err := os.Create(name)

defer func() {
    if err == nil {
        err = fd.Close()
    }
    
    if err != nil {
       _ = os.Remove(name)
    } else {
       fmt.Fprintf(gd.out, "Downloaded: %s\n", name)
    }
}()

if err != nil {
   return err
}
```